### DISCUSSION TITLE
Failed to search all shard leaders error

### DISCUSSION THREAD

**Discussion by dee-corr** [milvus-io/milvus#19377] [Q&A and General discussion]
Hello,

Could anyone help me understand the error below? It happens when I change nprobe to be anything other than 1.

<img width="452" alt="image" src="https://user-images.githubusercontent.com/51026617/191732263-5a022de9-f220-43ec-90ce-5df4afef9784.png">

The results in the screenshot represent the most similar to the searched vector, right? How do I get the most dissimilar vectors?

**Comment by yhmo** [comment]
The screenshot is blur. I saw the "Search params check failed" but I can't see the full text.
Did you create index on the field "morgan_fingerprint"?

**Reply by dee-corr** [reply]
Apologies, here is a better screenshot:

<img width="1667" alt="image" src="https://user-images.githubusercontent.com/51026617/192863412-e893d80d-8f3b-4c88-bed8-52758e956e30.png">

yes BIN_IVF_FLAT index on the morgan_fingerprint vector field

**Reply by yhmo** [reply]
This error is different to the first screenshot. This error indicates the querynode is down. What kind of deployment? standalone or cluster?
If your milvus service is cluster mode, please collect the log of querynode and sent to us. Then restart the querynode container and search again. If the search request still failed, try release the collection and load the collection and search again.

**Reply by dee-corr** [reply]
Hi @yhmo, apologies, there seems to be a few issues going on. I decided to redeploy on a fresh instance and reload the collections to see if that would help. It did but there is still a problem with creating indexes. This time, I have created the index using the Attu front end, but it was also not working with pymilvus. As you can see below it is in a constant loading state:

<img width="1631" alt="image" src="https://user-images.githubusercontent.com/51026617/193481134-bd335011-0fcf-4ec6-97b6-a27c81ab9a38.png">

These are the errors in the logs:
`milvus-standalone | [2022/10/02 23:27:54.094 +00:00] [ERROR] [indexcoord/index_builder.go:170] ["index builder peek client error, there is no available"] [stack="github.com/milvus-io/milvus/internal/indexcoord.(*indexBuilder).process\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_builder.go:170\ngithub.com/milvus-io/milvus/internal/indexcoord.(*indexBuilder).schedule\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_builder.go:152"]
milvus-standalone | [2022/10/02 23:27:54.094 +00:00] [ERROR] [indexcoord/index_builder.go:170] ["index builder peek client error, there is no available"] [stack="github.com/milvus-io/milvus/internal/indexcoord.(*indexBuilder).process\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_builder.go:170\ngithub.com/milvus-io/milvus/internal/indexcoord.(*indexBuilder).schedule\n\t/go/src/github.com/milvus-io/milvus/internal/indexcoord/index_builder.go:152"]
milvus-standalone | [2022/10/02 23:27:59.681 +00:00] [ERROR] [datacoord/services.go:807] ["failed to complete compaction"] [planID=436404927585845284] [error="plan 436404927585845284's state is 3"] [stack="github.com/milvus-io/milvus/internal/datacoord.(*Server).CompleteCompaction\n\t/go/src/github.com/milvus-io/milvus/internal/datacoord/services.go:807\ngithub.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).CompleteCompaction\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:296\ngithub.com/milvus-io/milvus/internal/proto/datapb._DataCoord_CompleteCompaction_Handler.func1\n\t/go/src/github.com/milvus-io/milvus/internal/proto/datapb/data_coord.pb.go:5133\ngithub.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.UnaryServerInterceptor.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/tracing/opentracing/server_interceptors.go:38\ngithub.com/milvus-io/milvus/internal/proto/datapb._DataCoord_CompleteCompaction_Handler\n\t/go/src/github.com/milvus-io/milvus/internal/proto/datapb/data_coord.pb.go:5135\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.44.0/server.go:1282\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.44.0/server.go:1616\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.44.0/server.go:921"]
milvus-standalone | [2022/10/02 23:27:59.681 +00:00] [ERROR] [datanode/compactor.go:524] ["complete compaction wrong"] [planID=436404927585845284] [reason="plan 436404927585845284's state is 3"] [stack="github.com/milvus-io/milvus/internal/datanode.(*compactionTask).compact\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compactor.go:524\ngithub.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:99"]`

**Reply by dee-corr** [reply]
There doesn't seem to be any issues from this docker resource usage point of view:
<img width="1631" alt="image" src="https://user-images.githubusercontent.com/51026617/193481377-ca94bfa8-0b91-49c4-91ec-2e7e9a014f59.png">

This is a standalone instance and have tried redeploying and releasing and reloading but this didn't fix the problem

**Reply by dee-corr** [reply]
@yhmo is there specific logs I can provide or anything else I can try?

**Reply by yhmo** [reply]
From the log, it seems the milvus you deployed is v2.0.2. If so, please upgrade to latest version, this might be a bug of old version.

**Reply by dee-corr** [reply]
It's actually v2.1

<img width="853" alt="image" src="https://user-images.githubusercontent.com/51026617/193870914-55d180f6-eb2a-49de-9b10-674eb3c17e0f.png">

Could it be a bug in v2.1?

**Reply by dee-corr** [reply]
I also tried version 2.1.4 but no luck

**Reply by yhmo** [reply]
I browse the source code, there are some modifications for compaction logic recently. The error you encountered is pretty like this issue: https://github.com/milvus-io/milvus/issues/18674, and there is a fix for this issue on 23th Aug: https://github.com/milvus-io/milvus/pull/18795, and this fix is in v2.1.2.
But you say the 2.1.4 also has this defect, I think we need the steps to reproduce it. Could you describe in details how you get this error?

**Reply by dee-corr** [reply]
Yes I will pass on my deployment config and my python script for loading if it will help to investigate the issue faster

**Reply by dee-corr** [reply]
Here is a test script:

```
import numpy as np
import pandas as pd
from pymilvus import (
    connections,
    utility,
    FieldSchema,
    CollectionSchema,
    DataType,
    Collection,
)
from bitstring import BitArray
from random import randint

fp_size = 8192

milvus_host = "<host here>"
milvus_port = "19530"

collection_name = "emols_fp_1024_test"


print("*** Connect to milvus ***")
connections.connect("default", host=milvus_host, port=milvus_port)

has = utility.has_collection(collection_name)
print(f"Does collection exist in Milvus: {has}")

if not has:
    print("*** Create collection ***")
    fields = [
        FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=True),
        FieldSchema(name="smiles", dtype=DataType.VARCHAR, max_length=200),
        FieldSchema(
            name="morgan_fingerprint",
            dtype=DataType.BINARY_VECTOR,
            dim=fp_size,
            is_primary=False,
        ),
    ]
    schema = CollectionSchema(fields, collection_name)
    ligands_collection = Collection(collection_name, schema, consistency_level="Strong")

    print("Collection created")
else:
    ligands_collection = Collection(collection_name)
    print(ligands_collection)

    
chunk_size = 100000
smiles = ["test_string"] * chunk_size
for i in range(0, 240):
    fps = [BitArray([randint(0, 1) for i in range(1024)]).bin] * chunk_size
    fps_bytes = list(map(lambda p: bytes(p, encoding='utf-8'), fps))
    ligands_collection.insert([smiles, fps_bytes])



print("*** Load collection ***")
ligands_collection.load()


print("*** Create index TANIMOTO BIN_IVF_FLAT ***")
index = {
    "index_type": "BIN_IVF_FLAT",
    "metric_type": "TANIMOTO",
    "params": {"nlist": 19596},
}

ligands_collection.create_index("morgan_fingerprint", index)

```

**Reply by yhmo** [reply]
I deploy a standalone v2.1.4, and run this script on my machine.
I use some utility methods to check the segments and index state:
```
info = utility.get_query_segment_info(collection_name=collection_name)
print(info)
print("segment count", len(info))

prog = utility.index_building_progress(collection_name=collection_name)
print(prog)
```

I observed something:
- it takes about 15 minutes to generate/insert 24M binary vectors with 8192 dim.
- it takes about 8 minutes to load the 24M vectors into memory
- there are 120 segments generated, and each segment has about 200000 rows.
- the index process is working but is very slow. building an index for a segment cost 15 minutes. seems the index process for binary vector is single-thread(only one CPU is busy).
- during index is building, search can work but is brute-force search.

**Reply by yhmo** [reply]
The `nlist` value 19596 is too large. We recommend the nlist to be `4*sqrt(n)`, the n is not the total row count, it is an estimated value of average row count of each segment. Since each segment has about 200000 rows, the nlist = 4*sqrt(200000) = 1787. Small nlist value can reduce the time to build index, so I change the nlist to 1024, each segment takes 4~5 minutes to build index, but is still slow.

I think the index algorithm for binary vector is not optimized. I already created an issue for the index/search engine: https://github.com/milvus-io/knowhere/issues/494

**Reply by yhmo** [reply]
Even changing the nlist to 1024, it still takes 10 hours to build index for all the segments. So, you will see the CREATE INDEX status keeps running for a long time. Search is available during the 10 hours.
